### PR TITLE
feat(platform): approval flow dataSource updates

### PR DESCRIPTION
**Pull request by dmitry-stepanenko**
## Related Issue(s)

closes https://github.com/SAP/fundamental-ngx/issues/6537

## Description
- Removed one unified dataSource in favour of common DataSource + DataProvider implementation per entity. `dataSource` input has been removed, `usersDataSource`, `watchersDataSource` and `teamsDataSource` were added
- approvalFlow is now being passed as `value ` input
- new output `valueChange` added. It emits whenever approvalFlow is changed (either watchers or list of approval flow nodes)
- Send reminders logic is no longer inside the approval flow. Whenever `Send Reminders` button is clicked, `sendReminders` output will emit. Notification about successful send of reminder should be shown manually by parent component.
- Updated `ApprovalTeam` interface: now it's expected to have the full list of team users instead of their ids
- Updated `ApprovalUser` interface: now it should have `teamId` property

There's an issue with selected watchers: due to usage of cloneDeep in examples, selected watchers are not displayed as selected in the list in edit mode. There's a separate issue for that https://github.com/SAP/fundamental-ngx/issues/7408

##### During Implementation

1. Visual Testing:

-   [x] visual misalignments/updates
-   [x] check Light/Dark/HCB/HCW themes
-   [x] RTL/LTR - proper rendering and labeling
-   [x] responsiveness(resize)
-   [x] Content Density (Cozy/Compact/(Condensed))
-   [x] States - hover/disabled/focused/active/on click/selected/selected hover/press state
-   [x] Interaction/Animation - open/close, expand/collapse, add/remove, check/uncheck
-   [x] Mouse vs. Keyboard support
-   [x] Text Truncation

2. API and functional correctness

-   [x] check for console logs (warnings, errors)
-   [x] API boundary values
-   [x] different combinations of components - free style
-   [x] change the API values during testing

3. Documentation and Example validations

-   [x] missing API documentation or it is not understandable
-   [x] poor examples
-   [n/a] Stackblitz works for all examples

4. Accessibility testing
5. Browser Testing - Edge, Safari, Chrome, Firefox

##### PR Quality

-   [x] the commit message(s) follows the guideline:
        https://github.com/SAP/fundamental-ngx/blob/main/CONTRIBUTING.md
-   [n/a] tests for the changes that have been done
-   [x] all items on the PR Review Checklist are addressed :
        https://github.com/SAP/fundamental-ngx/wiki/PR-Review-Checklist
-   [n/a] Run npm run build-pack-library and test in external application
-   [n/a] update `README.md`
-   [x] [Breaking Changes Wiki](https://github.com/SAP/fundamental-ngx/wiki/Breaking-Changes)

### CODE REVIEW

File: libs/platform/src/lib/approval-flow/interfaces/approval-data-source.ts
Comment (fkolar) [inline libs/platform/src/lib/approval-flow/interfaces/approval-data-source.ts:7]: My issue initially with this is that it does not implement DataSource (open, close) interface just like the other data sources.  I would reimplement this class. There needs to be a way way to listen for data changes which AF component listen. 

Or maybe two since we have actual two source teams, users. I would not try to adjust existing implementaion

File: libs/platform/src/lib/approval-flow/helpers/data-source.ts
Comment (fkolar) [inline libs/platform/src/lib/approval-flow/helpers/data-source.ts:10]: This introduce allot of complexity. 

AF Component will have one way of listening for data coming in from provider. This is why we have open() method in the DataSource Interface. 

I know that this AFDS requires two data sources, but we should follow the same semantics like other implementations.

**Review by fkolar — CHANGES_REQUESTED** [review]

File: apps/docs/src/app/platform/component-docs/platform-approval-flow/platform-approval-flow-examples/platform-approval-flow-example.component.html
Comment (fkolar) [inline apps/docs/src/app/platform/component-docs/platform-approval-flow/platform-approval-flow-examples/platform-approval-flow-example.component.html:5]: User not users, userDataSource  - Data Source for a User Type

File: apps/docs/src/app/platform/component-docs/platform-approval-flow/platform-approval-flow-examples/platform-approval-flow-example.component.html
Comment (fkolar) [inline apps/docs/src/app/platform/component-docs/platform-approval-flow/platform-approval-flow-examples/platform-approval-flow-example.component.html:3]: Let's rename this either to `approvalProces` or better just `value`

this component have one input VALUE and 3 sources.

File: apps/docs/src/app/platform/component-docs/platform-approval-flow/platform-approval-flow-examples/platform-approval-flow-example.component.html
Comment (fkolar) [inline apps/docs/src/app/platform/component-docs/platform-approval-flow/platform-approval-flow-examples/platform-approval-flow-example.component.html:16]: rename to `nodeUpdate` or `valueChange` so we can keep it similar to the other components

File: libs/platform/src/lib/approval-flow/approval-flow-add-node/approval-flow-add-node.component.ts
Comment (fkolar) [inline libs/platform/src/lib/approval-flow/approval-flow-add-node/approval-flow-add-node.component.ts:168]: Just a question, what does this .match do exactly.

**Review by fkolar — CHANGES_REQUESTED** [review]

File: apps/docs/src/app/platform/component-docs/platform-approval-flow/platform-approval-flow-examples/platform-approval-flow-example.component.html
Comment (dmitry-stepanenko) [reply apps/docs/src/app/platform/component-docs/platform-approval-flow/platform-approval-flow-examples/platform-approval-flow-example.component.html:3]: done

File: apps/docs/src/app/platform/component-docs/platform-approval-flow/platform-approval-flow-examples/platform-approval-flow-example.component.html
Comment (dmitry-stepanenko) [reply apps/docs/src/app/platform/component-docs/platform-approval-flow/platform-approval-flow-examples/platform-approval-flow-example.component.html:16]: done

File: apps/docs/src/app/platform/component-docs/platform-approval-flow/platform-approval-flow-examples/platform-approval-flow-example.component.html
Comment (dmitry-stepanenko) [reply apps/docs/src/app/platform/component-docs/platform-approval-flow/platform-approval-flow-examples/platform-approval-flow-example.component.html:5]: done

File: libs/platform/src/lib/approval-flow/approval-flow-add-node/approval-flow-add-node.component.ts
Comment (dmitry-stepanenko) [reply libs/platform/src/lib/approval-flow/approval-flow-add-node/approval-flow-add-node.component.ts]: This is needed to trigger initial loading of data without any filters applied. I've made `match`'s function param optional, so it looks slightly better. Also added a comment explaining why this is needed

File: libs/platform/src/lib/approval-flow/approval-flow-add-node/approval-flow-add-node.component.ts
Comment (fkolar) [inline libs/platform/src/lib/approval-flow/approval-flow-add-node/approval-flow-add-node.component.ts:168]: This is better. instead of passing newMap

**Review by fkolar — APPROVED** [review]
Good Job. 
Let's make sure that Approval flow is released under breaking changes. Please see local md documents to learn more about this.

**Comment by dmitry-stepanenko** [comment]
> Good Job. Let's make sure that Approval flow is released under breaking changes. Please see local md documents to learn more about this.

Yes, I'll update breaking changes once this PR is ready to be merged. Currently it's blocked by #7408 ticket, I'm working in it

**Review by Betrozov — APPROVED** [review]

File: libs/platform/src/lib/approval-flow/approval-flow.component.ts
Comment (dimamarksman) [inline libs/platform/src/lib/approval-flow/approval-flow.component.ts:247]: why do we call it `approvalState` if it returns `ApprovalProcess`? Minor but looks a little bit weird

File: libs/platform/src/lib/approval-flow/approval-flow.component.ts
Comment (dimamarksman) [inline libs/platform/src/lib/approval-flow/approval-flow.component.ts:480]: in some places I see dataSource.close() approach but here we don't.
Is it intentional?

**Review by dimamarksman — COMMENTED** [review]
Looks good!
left a few questions

File: apps/docs/src/app/platform/component-docs/platform-approval-flow/platform-approval-flow-examples/platform-approval-flow-example.component.ts
Comment (dmitry-stepanenko) [reply apps/docs/src/app/platform/component-docs/platform-approval-flow/platform-approval-flow-examples/platform-approval-flow-example.component.ts:18]: I'm following the existing structure, where we have all data sources in `platform/shared`

File: libs/platform/src/lib/approval-flow/approval-flow.component.ts
Comment (dmitry-stepanenko) [reply libs/platform/src/lib/approval-flow/approval-flow.component.ts:130]: Again, I'm following the existing approaches. E.g. for combobox it's called in the same way. I'll rename it and other inputs of AF to the form of `usersDataProviderEntityKey`

File: libs/platform/src/lib/approval-flow/approval-flow.component.ts
Comment (dmitry-stepanenko) [reply libs/platform/src/lib/approval-flow/approval-flow.component.ts:247]: Yeah, I agree, will rename

File: apps/docs/src/app/platform/component-docs/platform-approval-flow/platform-approval-flow-examples/platform-approval-flow-example.component.ts
Comment (dmitry-stepanenko) [reply apps/docs/src/app/platform/component-docs/platform-approval-flow/platform-approval-flow-examples/platform-approval-flow-example.component.ts]: fixed

File: libs/platform/src/lib/approval-flow/approval-flow.component.ts
Comment (dmitry-stepanenko) [reply libs/platform/src/lib/approval-flow/approval-flow.component.ts]: done

File: libs/platform/src/lib/approval-flow/approval-flow.component.ts
Comment (dmitry-stepanenko) [reply libs/platform/src/lib/approval-flow/approval-flow.component.ts]: done

File: libs/platform/src/lib/approval-flow/approval-flow.component.ts
Comment (dmitry-stepanenko) [reply libs/platform/src/lib/approval-flow/approval-flow.component.ts:493]: missed this, thanks. added now

File: apps/docs/src/app/platform/component-docs/platform-approval-flow/platform-approval-flow-examples/platform-approval-flow-example.component.ts
Comment (dimamarksman) [reply apps/docs/src/app/platform/component-docs/platform-approval-flow/platform-approval-flow-examples/platform-approval-flow-example.component.ts:19]: I see, ok
Would be great to keep specific things in the scope of a specific module.
Created a separate issue to track it https://github.com/SAP/fundamental-ngx/issues/7501

File: libs/platform/src/lib/approval-flow/approval-flow-add-node/approval-flow-add-node.component.ts
Comment (dimamarksman) [reply libs/platform/src/lib/approval-flow/approval-flow-add-node/approval-flow-add-node.component.ts:221]: IMHO if we offer the ApprovalFlowDataSource in the doc as one ready to go and actually are using it in the approval-flow component we need to have this method implemented so it does the job.
Otherwise it's not clear and a user might be confused.
Could you please add a minimal implementation for `dataSource.close()` method so at least it closes a stream that was opened via `dataSource.open()`. I suppose this change should be done in ComboboxDataSource class.

Comment (dmitry-stepanenko) [reply libs/platform/src/lib/approval-flow/approval-flow-add-node/approval-flow-add-node.component.ts:221]: sure, done

Comment (dimamarksman) [reply libs/platform/src/lib/approval-flow/approval-flow-add-node/approval-flow-add-node.component.ts:221]: thank you

**Review by dimamarksman — APPROVED** [review]

**Pull request merged**
